src/tiledb/cloud/bioimg/ingestion.py

```diff
@@ -8,10 +8,185 @@
+        for input, output in io_uris:
+            with vfs.open(input) as src:
+                OMETiffConverter.to_tiledb(src, output, **kwargs)
+
+            if embedding_model:
+                # Create the embeddings output paths
+                embeddings_flat_uri, embeddings_ivf_flat_uri = get_embeddings_uris(
+                    output
+                )
+
+                # Split image [level=arg] in rectangular patches [arg] size
+                def split_in_patches(
+                    image: np.ndarray,
+                    embedding_level: int,
+                    embedding_grid: Tuple[int, int],
+                ) -> Sequence[np.ndarray]:
+                    level_shape_w, level_shape_h = image.level_dimensions[
+                        embedding_level
+                    ]
+                    # Calculate the region size based on the desired
+                    # number of rows and columns
+                    grid_row_num, grid_col_num = embedding_grid
+                    region_height = level_shape_h // grid_row_num
+                    region_width = level_shape_w // grid_col_num
+                    # Loop through the image and extract each region
+                    patches = []
+                    for i in range(grid_row_num):
+                        for j in range(grid_col_num):
+                            location = (i, j)
+                            patches.append(
+                                image.read_region(
+                                    location,
+                                    embedding_level,
+                                    (region_height, region_width),
+                                )
+                            )
+                    return patches
+
+                with TileDBOpenSlide(output) as image:
+                    # Filter out patches with [arg] % non-blank coverage
+                    filtered = split_in_patches(image, embedding_level, embedding_grid)
+
+                    patches_array = np.array([])
+                    for patch in filtered:
+                        patch_transformed = patch[np.newaxis]
+                        if patches_array.any():
+                            patches_array = np.concatenate(
+                                (patches_array, patch_transformed), axis=0
+                            )
+                        else:
+                            patches_array = patch_transformed
+
+                    # Create embedding of the image given its patches
+                    embeddings = calculate_model(patches_array, embedding_model)
+                    # Create flat index of the embedding for faster vector search
+                    tmp_features_file = "/tmp/features"
+                    with open(tmp_features_file, "wb") as f:
+                        np.array(embeddings.shape, dtype="uint32").tofile(f)
+                        np.array(embeddings).astype("float32").tofile(f)
+
+                    vs.ingest(
```

Comment:
Can you explain a bit the use case that you want to support?

From what I understand this is creating one vector index per image, storing the embedding of different patches of the image. Does this mean that you want to find similar patches within the image? Do you want to do any cross image queries?

Reply:
The idea here is to find similar patches across multiple images. We don't have any specific customer request for any of that. The idea behind the patches though is that in this data environment is quite unlikely to find similarities across whole images. Each image is quite different from one another as a whole based on the depicted cell. Also from my knowledge due to the blank space outside the cells captured by sensor it is quite possible to introduce a lot of noise in the model if fed to it as a whole. So my thought was that a future user would like to search for similarities across images of a specific region of the query image, or in a next step to be able to query a region from the viewer and find similar "abnormalities" - that he can justify and I can't- across the images.